src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs

```diff
@@ -167,22 +179,63 @@ public TypeSymbolWithAnnotations AsNullableReferenceOrValueType(CSharpCompilatio
         /// </summary>
         public abstract TypeSymbolWithAnnotations AsNullableReferenceType();
         public abstract TypeSymbolWithAnnotations AsNotNullableReferenceType();
-        public abstract TypeSymbolWithAnnotations AsObliviousReferenceType();
 
         public abstract TypeSymbolWithAnnotations WithModifiers(ImmutableArray<CustomModifier> customModifiers);
+        protected abstract TypeSymbolWithAnnotations WithNonNullTypes(bool nonNullTypes);
 
         public abstract TypeSymbol TypeSymbol { get; }
         public virtual TypeSymbol NullableUnderlyingTypeOrSelf => TypeSymbol.StrippedType();
 
         /// <summary>
-        /// Is this a nullable reference or value type.
-        /// If it is a nullable value type, <see cref="TypeSymbol"/>
+        /// Returns:
+        /// true if this is a nullable reference or value type;
+        /// false if this is an unannotated reference type and [NonNullTypes(true)],
+        /// or a value type regardless of [NonNullTypes]; and
+        /// null if an unannotated reference type and [NonNullTypes(false)].
+        /// If this is a nullable value type, <see cref="TypeSymbol"/>
         /// returns symbol for constructed System.Nullable`1 type.
-        /// If it is a nullable reference type, <see cref="TypeSymbol"/>
+        /// If this is a nullable reference type, <see cref="TypeSymbol"/>
         /// simply returns a symbol for the reference type.
         /// </summary>
         public abstract bool? IsNullable { get; }
 
+        /// <summary>
+        /// Returns:
+        /// false for string, int, T;
+        /// true for string?, T? where T : class; and
+        /// true for int?, T? where T : struct.
+        /// </summary>
+        public abstract bool IsAnnotated { get; }
+
+        /// <summary>
+        /// [NonNullTypes] context used for determining
+        /// whether unannotated types are not nullable.
+        /// </summary>
+        public abstract bool NonNullTypes { get; }
+
+        /// <summary>
+        /// Returns:
+        /// true if annotated;
+        /// false if unannotated and [NonNullTypes(true); and
+        /// null if unannotated and [NonNullTypes(false).
+        /// </summary>
+        /// <remarks>
+        /// This property considers IsAnnotated and NonNullTypes only. Compare with
+        /// IsNullable that also considers IsValueType. (Specifically, IsNullable==false
+        /// for an unannotated value type, regardless of [NonNullTypes].)
+        /// </remarks>
+        internal bool? IsAnnotatedWithNonNullTypesContext
```

Comment:
I'd consider making this private, and adjusting the tests accordingly (verify IsAnnotated and NonNullTypes separately, or replicate logic?).